components/script/dom/blob.rs

```diff
@@ -222,6 +229,42 @@ impl BlobMethods for Blob {
         let blob_impl = BlobImpl::new_sliced(rel_pos, self.blob_id.clone(), type_string);
         Blob::new(&*self.global(), blob_impl)
     }
+
+    // https://w3c.github.io/FileAPI/#text-method-algo
+    fn Text(&self) -> Rc<Promise> {
+        let in_compartment_proof = AlreadyInCompartment::assert(&self.global());
+        let p = Promise::new_in_current_compartment(
+            &self.global(),
+            InCompartment::Already(&in_compartment_proof),
+        );
+        let bytes = self.get_bytes().expect("Could not get blob contents");
```

Comment:
The next step here will be to make read_file_async take a function argument (`Box<Fn(Rc<Promise>, Result<Vec<u8>, ())>`) so that we can call the function from the thread spawned in read_file_async and pass the retrieved bytes and the rooted promise. This will allow us to call read_file_async from both Text and ArrayBuffer and pass a closure (`|promise, bytes| { ... }`) that does the UTF-8 decoding or array buffer creation and resolves the promise.

Reply:
How do I pass the SafeJsContext object (line 256) in the closure?

Reply:
Ah, great question. You can use `promise.global()` to get a GlobalScope, and call `get_cx()` on that to get a SafeJsContext.

Reply:
I am getting these errors again. I cannot figure out which part of the code is causing this problem.
```
error[E0277]: `dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)` cannot be sent between threads safely
    --> components/script/dom/globalscope.rs:1259:33
     |
1259 |             let _ = task_source.queue_with_canceller(
     |                                 ^^^^^^^^^^^^^^^^^^^^ `dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)` cannot be sent between threads safely
     |
     = help: the trait `std::marker::Send` is not implemented for `dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)`
     = note: required because of the requirements on the impl of `std::marker::Send` for `std::ptr::Unique<dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)>`
     = note: required because it appears within the type `std::boxed::Box<dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)>`
     = note: required because it appears within the type `[closure@components/script/task.rs:27:15: 27:19 trusted_promise:dom::bindings::refcounted::TrustedPromise, f:std::boxed::Box<dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)>, bytes:std::result::Result<std::vec::Vec<u8>, ()>]`
     = note: required because of the requirements on the impl of `task::TaskOnce` for `dom::globalscope::GlobalScope::read_file_async::{{closure}}#0::resolve_promise<[closure@components/script/task.rs:27:15: 27:19 trusted_promise:dom::bindings::refcounted::TrustedPromise, f:std::boxed::Box<dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)>, bytes:std::result::Result<std::vec::Vec<u8>, ()>]>`

error[E0277]: `dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)` cannot be sent between threads safely
    --> components/script/dom/globalscope.rs:1257:9
     |
1257 |         thread::spawn(move || {
     |         ^^^^^^^^^^^^^ `dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)` cannot be sent between threads safely
     |
     = help: the trait `std::marker::Send` is not implemented for `dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)`
     = note: required because of the requirements on the impl of `std::marker::Send` for `std::ptr::Unique<dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)>`
     = note: required because it appears within the type `std::boxed::Box<dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)>`
     = note: required because it appears within the type `[closure@components/script/dom/globalscope.rs:1257:23: 1266:10 recv:profile_traits::ipc::IpcReceiver<std::result::Result<net_traits::filemanager_thread::ReadFileProgress, net_traits::filemanager_thread::FileManagerThreadError>>, task_source:task_source::file_reading::FileReadingTaskSource, trusted_promise:dom::bindings::refcounted::TrustedPromise, f:std::boxed::Box<dyn std::ops::Fn(std::rc::Rc<dom::promise::Promise>, std::result::Result<std::vec::Vec<u8>, ()>)>, canceller:task::TaskCanceller]`
```

Reply:
You're right! I played around with it on the rust playground, and realized that we need to attach `+ Send` to the Fn trait type, like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=61b19f1f7680e49a2c8b960327d16156).

Reply:
I found something quite similar in the official Rust book, in [this section](https://doc.rust-lang.org/book/ch20-02-multithreaded.html#building-the-threadpool-struct-using-compiler-driven-development).

>The F type parameter also has the trait bound Send and the lifetime bound 'static, which are useful in our situation: we need Send to transfer the closure from one thread to another and 'static because we don’t know how long the thread will take to execute.

Why don't we need lifetime bound `'static` in our code?

Reply:
'static is the default lifetime bound if none is specified.